File: src/main/java/io/protobase/f7/formulas/logic/IFERROR.java
package io.protobase.f7.formulas.logic;

import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class IFERROR extends AbstractFormula {
  public static FormulaName NAME = FormulaName.IFERROR;
  public static IFERROR SELF = new IFERROR();

  public IFERROR() {
    super();
  }

  public IFERROR(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 2, NAME);
    Object value = Converters.first(collateralLookup.apply(origin, values[0]));
    return value instanceof F7Exception ? Converters.first(collateralLookup.apply(origin, values[1])) : value;
  }
}

File: src/main/js/formulas/logic/IFERROR.ts
import { F7Exception } from "../../errors/F7Exception";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class IFERROR extends AbstractFormula {
  static SELF: IFERROR = new IFERROR();
  NAME = FormulaName.IFERROR;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 2, this.NAME);
    const value = this.collateralLookup(origin, values[0]);
    if (Converters.first(value) instanceof F7Exception) {
      return this.collateralLookup(origin, values[1]);
    }
    return value;
  }
}
